### DISCUSSION TITLE
rust-course/first-try/hello-world

### DISCUSSION THREAD

**Discussion by sunface** [sunface/rust-course#625] [章节评论区]
https://course.rs/first-try/hello-world.html

**Comment by ZyMousai** [comment]
这本书学着学着我就笑起来了，太幽默了。作者大大厉害

**Reply by overlords** [reply]
哈哈哈！
螃蟹学院的搞笑校长

**Comment by sunface** [comment]
> 这本书学着学着我就笑起来了，太幽默了。作者大大厉害

那你在后续章节得小心大话西游效应：随着你越来越成熟，看大话西游时就越来越想哭 :D

**Reply by jiechen257** [reply]
笑死 :P

**Comment by DeWjjj** [comment]
我单方面宣布，这一节不用看。
太离谱了，还没教语法呢，就给我：：起来了。

**Reply by sunface** [reply]
从入门到放弃

**Reply by mxismean** [reply]
门都得好多遍才能入😂

**Reply by caoxiemeihao** [reply]
那开始不劝退一波不符合 Rust 尿性 🙃

**Reply by mikuwife** [reply]
在门槛上反复横跳

**Reply by zhaoyuanmeng** [reply]
秀我一脸

**Reply by mikuwife** [reply]
从入门到入门再到入门

**Reply by 478619644h** [reply]
三过家门而不入

**Reply by 0xlauu** [reply]
电脑已砸，就这篇就换了一台电脑💻

**Reply by XSAXL818** [reply]
开头都说初印象了，就是看看代码长啥样

**Comment by zengleyi** [comment]
非常棒，比官方版本的更容易入门，赞！

**Comment by jyutjyu** [comment]
谢谢分享。

**Comment by ra9fael** [comment]
great daze!

**Comment by sweet2honey** [comment]
看到你的骚话，感到学编程更快乐了。这次的rust肯定要好好学了（之前看过的都忘了

**Comment by kaixinbaba** [comment]
> 还有一个相亲，啊呸

我好像已经知道作者的为人了。。doge

**Comment by findre** [comment]
看到 for index, item in Listobj.enumerate() 梦回 Python3 的元组拆包和enumerate组合。
看到 for region in regions 可用，梦回Python3 元编程。
看到 .map(|field| field.trim()) 蒙圈了，看到 凭空冒出来的 debug_assertions 准备退出教程。
看到 作者也颤抖我又坐了回来，看来不是我一个在蒙圈

**Reply by ZhongFuze** [reply]
enumerate() iter() for x in xx .map(lambda x, x.split(","))
妈耶, 和你的心路历程一样hh

**Reply by AllenWu233** [reply]
Python爱好者表示一模一样233

**Comment by yurnery** [comment]
单方向宣布, 比官方版本的好多了, 语言幽默, 也不会让你觉得太过基础, 也不会太难!

**Reply by michael1735** [reply]
我也觉得，我官方的看了一半还是不懂到底怎么写程序才能让编译器通过

**Comment by E-Jaeger** [comment]
不喜欢这种写作风格，废话太多

**Reply by sunface** [reply]
不喜欢这种评论风格，酸不溜秋

**Reply by ColinZeb** [reply]
哈哈 喜欢高等数学同济版

**Reply by Aemoz** [reply]
同济版～  惨痛的回忆

**Reply by Cateners** [reply]
我也不太喜欢这种风格，倒不是说废话太多，只是一路读过来总觉得别扭。
当然作者用爱发电还是要支持>_<

**Reply by ideazhao** [reply]
这已经算废话很少的了，你没见过真废话多的，一句话能解释的来一段话打比方，结果还没解释清楚，那才叫痛苦，我说的就是某前端大神~

**Comment by tabris233** [comment]
喜欢这种写做风格，不枯燥。

**Comment by DrAugus** [comment]
笑着笑着就哭了

**Reply by sunface** [reply]
我怀疑你看过大话西游

**Reply by DrAugus** [reply]
应该没人没看过吧

**Comment by shilohooo** [comment]
```Rust
records.map(|field| field.trim());
```
是否类似于Java的如下代码呢？
```Java
records.map(field -> field.trim());
```

**Reply by DreamSaddle** [reply]
哈哈，我觉得是这样的。

**Reply by DreamSaddle** [reply]
都是换汤不换药吧。

**Reply by ramonli** [reply]
翻页的箭头能不能放到正文后面，每次都要拖到最下面才能翻页

**Reply by sunface** [reply]
好的，建议收到，其实还可以用左右箭头🙂

**Reply by zzcan** [reply]
也形如 JS 的 records.map(field => field.trim());  :)

**Comment by colorbeach** [comment]
作者这么可爱，一定是蓝孩子吧😍

**Reply by gotoorder** [reply]
你的表情丐丐的

**Comment by nibaijing** [comment]
map(|field| field.trim())   |field|,两竖杆这是什么骚操作？

**Reply by sunface** [reply]
closure 闭包

**Reply by james-wangx** [reply]
```rust
let fields: Vec<_> = record
       .split(',')
       .map(|field| field.trim())
       .collect();
```
这段代码我一开始也是竖着写的，但是一格式化就变成这样横着了什么情况？可以调整吗？我想竖着。
```rust
let fields: Vec<_> = record.split(',').map(|field| field.trim()).collect();
```

**Reply by WvvvWv** [reply]
回复james-wangx，好像是rust格式化导致的，如果你是用的idea就去设置->语言和框架->Rust->Rustfmt里将两个框取消勾选。

**Comment by bbmeme05** [comment]
![img](https://s3.bmp.ovh/imgs/2022/07/11/62681e9bdef29eb4.png)
卡在这里半个小时了。。。。

**Reply by bbmeme05** [reply]
删了package cache。更换 cargo清华源后好了

**Reply by bbmeme05** [reply]
我去。。。我应该先看下一章

**Reply by sunface** [reply]
哈哈，我刚准备说的，没想到你自己发现了

**Comment by moruifei** [comment]
mark, 差点没看懂。big old有点跳了

**Comment by 2014** [comment]
学习曲线被楼主的骚话抹平了一半

**Comment by ixqbar** [comment]
let penguin_data = "\  //这里斜杠是麻作用？

**Comment by CenBoMin** [comment]
原来突然头晕就是这种感觉啊！

**Comment by 865881900** [comment]
if cfg!(debug_assertions) {
      // 输出到标准错误输出
      eprintln!("debug: {:?} -> {:?}",
                record, fields);
    }

看到这一行代码, 我就知道该到了放弃的时候了
遥想当初意气风发, 到如今举步维艰
哎, 真是命运多舛, 此处不留爷,必有留爷处, 在下去也
对了, PHP是最好的语言, 没有之一

**Comment by jxiaof** [comment]
犹如一杯红茶(因为我喜欢红茶)

**Comment by bingoku** [comment]
初学者表示，这章比官方的第一个例子，"猜数字"还残暴。

**Comment by GoOutSide66** [comment]
00语言是啥语言🐶

**Reply by cloudship** [reply]
Object Oriented. 面向对象的编程

**Reply by amandafanny** [reply]
我还以为是一门新语言呢，没有搜到

**Comment by hanbin92381** [comment]
写的真有意思！越看越来劲😄

**Comment by yammyxing** [comment]
第n次入门，冲～

**Comment by ismart-yuxi** [comment]
例子有点差强人意，要重点突出mut

**Comment by HiJohnnyLee** [comment]
很棒，适合摸鱼的时候看

**Comment by kuuyee** [comment]
风趣幽默！ 真棒

**Comment by Createitv** [comment]
看着看着就笑了，敲着敲着就哭了

**Reply by BiteFoo** [reply]
我来看看热闹的

**Comment by zhongxiao37** [comment]
纯好奇，Ruby里面有用”!“解构操作么？

**Comment by amandafanny** [comment]
OO 语言是哪个语言？

**Reply by ghost** [reply]
指的面向对象吧

**Comment by Kori000** [comment]
这语法, 亏贼~

**Comment by Claude-Zq** [comment]
真生动😋

**Comment by hexWars** [comment]
`是的，在 认识 Cargo中，曾经介绍过 --release 参数`
没有吧...我没看到

**Comment by xiamuguizhi** [comment]
转载的: 


这段 Rust 代码定义了一个函数 greet_world，该函数包含了三个字符串变量以及一个字符串数组。其中，每个字符串变量代表不同语言中的问候语，分别为：

southern_germany: 南德意志方言中常用的问候语“Grüß Gott!”。
chinese: 中文中常用的问候语“世界，你好”。
english: 英语中常用的问候语“World, hello”。
而字符串数组 regions 包含了这三个字符串变量，并在循环中依次遍历每个元素，将其引用（即传递其内存地址）传递给 println! 宏进行输出。循环过程中，region 变量分别表示数组中的每个元素。使用 iter() 方法可以返回一个迭代器，它允许你对数组的元素进行遍历。

最终，在 main 函数中调用 greet_world 函数，打印出了每种语言的问候语。

需要注意的是，在输出字符串时，还要使用 {} 占位符，它会被实际参数所替换。在这里，&region 表示引用 region 变量所代表的字符串。实际上，这个引用是不必要的，因为 Rust 会自动进行引用转换（deref coercion）。因此，也可以直接使用 println!("{}", region) 进行输出。

**Comment by blkcor** [comment]
mark! 2023.4.10

**Comment by JameFu** [comment]
5年+老码农mark

**Comment by shouldhavedone** [comment]
打卡纪念一下

**Comment by abnereel** [comment]
还好刚刚学过一点点基础语法，要不然懵逼了

**Comment by well-he** [comment]
看了好多次都这里都被劝退。。

**Comment by lm0303** [comment]
不懂就，oo语言是个啥，面相对象吗

**Reply by well-he** [reply]
OO(Object Oriented) 面向对象

**Comment by Viktor1813** [comment]
沃日，这一下子就有点难受啊

**Comment by jbeetle** [comment]
学过c/c++和JavaScript的同学很容易理解这些，其实语言发明者就把指针语言和脚本语言混搭在一起了，
我一时无法用一句话来总结它，但是懂的自然懂。知道这一点rust其实很简单

**Comment by thiszwinter** [comment]
up写的太好了 ， 赞

**Comment by hansenz42** [comment]
好棒，很喜欢这本书的行文风格，加油！

**Comment by lifei6671** [comment]
好神奇的语言，第一次感觉一门编程语言这么晦涩难懂。

**Comment by wangrongding** [comment]
好好好，上来就 `::` 是吧

**Comment by AiHaoChen** [comment]
对于会java的人来说，::、debug_assertions和Ok(length)是什么鬼，凭空出现的

**Reply by xiamuguizhi** [reply]
:: 是 Rust 中的路径分隔符。它用于访问一个模块中的函数、方法或常量。在 Rust 中，函数和方法也可以作为值进行传递和操作，而 :: 允许你引用和使用这些函数和方法。

debug_assertions 是一个 Rust 内置的全局变量，它是一个布尔值，用于判断是否处于调试模式。在 Rust 中，使用 debug_assertions 可以在调试模式下执行断言，而在发布模式下不执行。这样可以在调试时捕获一些错误，并在发布时避免性能损失。

Ok(length) 是 Rust 中的 Result 枚举类型的一个变体。Result 类型用于处理可能会发生错误的操作结果。Ok 是 Result 枚举的一个成员，表示操作成功，并包含一个关联的值（在这个例子中是 length）。Ok(length) 表示一个成功的操作结果，并将 length 作为成功的值返回。

@gtp3

**Comment by zjingwang** [comment]
看到|filed| field.trim()时，我傻掉了

**Comment by xupengfe** [comment]
if i == 0 || record.trim().len() == 0 { 这行可以去掉 i == 0 改为

因为 common name 那一行实际invalid 不会打印， 实际循环是从i:0 开始，避免理解错误， 只是教下大家  ||  用法也好。 所以可以加上注释行， 自行使用测试。
// if i == 0 || record.trim().len() == 0
if record.trim().len() == 0 {

**Comment by A-caibird** [comment]
这本书学着学着我就笑起来了，太幽默了。作者大大厉害

**Comment by inwind1** [comment]
“上面的 So Easy 的余音仍在绕梁” = “遥遥领先”

**Reply by BiteFoo** [reply]
已经自主可控，全面领先

**Comment by linan724** [comment]
大致能看懂；貌似是一个字符串数组切割然后去空格 遍历 ；大概是这个意思 ； 没具体看

**Comment by TaoContent** [comment]
已经有难度了，用法有点抽象

**Comment by CAESIUS-TIM** [comment]
代码部分可运行，但是是只读的，如果可以自己修改+运行就好了

**Comment by pzc1350** [comment]
nice

**Comment by oohy** [comment]
hello word 还有 :: 不合适  哈哈

**Comment by jmuchch** [comment]
看起来和JAVA区别不大，这就是lambda表达式，数组循环，类型转换，加上泛型推断

**Comment by tiantiven** [comment]
```rust
println!("{}", &region);
```
这里为什么要用引用？我把取地址运算符删掉也可以运行。
我是不是又钻牛角尖了？

**Reply by xiamuguizhi** [reply]
在 Rust 中，当你使用 `println!` 宏输出变量时，通常会使用引用（`&`）来避免所有权的转移。Rust 的所有权规则是这样的：当一个值被另一个变量使用或者作为函数的参数时，它的所有权就会被移动（move），原来的变量就不能再使用这个值了。

然而，`println!` 宏并不需要取得值的所有权。它只需要读取值，所以你可以传递一个对值的引用，这样原来的变量还可以继续使用这个值。

但是，对于实现了 `Copy` trait 的类型，比如整数、浮点数、布尔值、字符以及它们的引用等，当它们作为函数参数时，会进行复制（copy），而不是移动（move）。所以对这些类型的变量，你可以直接传递给 `println!` 宏，也可以传递它们的引用。

所以，你发现把取地址运算符删掉也可以运行，这是因为你输出的类型实现了 `Copy` trait。但是对于没有实现 `Copy` trait 的类型，比如 String 或者自定义的结构体，就必须使用引用，否则会因为所有权问题而编译失败。

至于你说的“钻牛角尖”，其实这并不是。理解和掌握 Rust 的所有权系统是学习 Rust 的重要部分，这对于写出安全和高效的 Rust 代码是非常必要的。这种深入探究的态度是非常好的！

@ GTP 4

**Reply by GH-jia** [reply]
可是，iter()迭代器返回的不就是原先集合元素的引用吗？代码中数组内是三个&str类型元素，那么迭代器返回的region的类型应该是&&str。或者说是我有些地方理解错了？

**Comment by raotaohub** [comment]
兄弟们👬🏻，用Edge浏览器，然后开大声朗读，口音选择 Chinese Xiao bei Online ，速度调中等，为本次再添乐趣，不用谢我！给我点赞😄

**Reply by BiteFoo** [reply]
好笑吗🤭

**Comment by yjshi2015** [comment]
看完之后，没有激动的心了，只有颤抖的手😭

**Comment by xyfll7** [comment]
王婆卖瓜，自卖自夸

**Comment by jackwangwj** [comment]
nice,成功run起来

**Comment by MatrixMachine** [comment]
|field| field.trim() 把我看傻了，不看解释我只能理解为管道中的|field|

**Comment by cybertheye** [comment]
这本书这个网站的搭建是rust搭的吗？

**Comment by sanrentai** [comment]
不知道是哪个插件导致的，出现了很多 : &str

**Reply by Beiun** [reply]
同样问题。刚开始不懂语法，还以为就是代码的内容。复制出去问gpt，发现压根没有，原来是解释说明。

**Comment by Skyfacon** [comment]
学完了真棒

**Comment by Skyfacon** [comment]
这么有意思的书，有木有出版啊，想买纸质支持下

**Comment by Endermanbugzjfc** [comment]
开头看你用了 `\ + 分行` 之余又 `if i == 0` 以为是没注意才多加了，直到后面有个 `{}cm` 才懂原来是要把表格标题去掉。 感觉自己大脑有点卡，不过还是建议再添行注释。 😅 :shipit:

**Comment by maobenqinag** [comment]
record .split(',') .map(|field| field.trim()) .collect();
scala的感觉

**Comment by IlIIIIIIlI** [comment]
学习

**Comment by jx-nan** [comment]
方法语法：由于 Rust 没有继承，因此 Rust 不是传统意义上的面向对象语言，但是它却从 OO 语言那里偷师了方法的使用 record.trim()，record.split(',') 等。

这里面的OO 是 GO 吗

**Reply by lilhammerfun** [reply]
应该不是在说go吧，可能是java？
go的`Yes and no.`还是蛮经典的吧
![image](https://github.com/sunface/rust-course/assets/113404700/d9b8a253-f001-4175-b94d-feaa0b65f5b3)

**Reply by jx-nan** [reply]
噢噢，谢谢，没写过go

**Reply by jshuaishuai** [reply]
OO 是面向对象

**Comment by gq-tang** [comment]
发现一个小bug:
>
> 对于 println 来说，我们没有使用其它语言惯用的 %s、%d 来做输出占位符，而是使用 {}，因为 Rust 在底层帮 我们做了大量工作，会自动识别输出数据的类型，例如当前例子，会识别为 ~~String~~ 类型。

应该是 会识别为&str类型吧

**Comment by winsonet** [comment]
说真的，我开始并未对 Rust 多大兴趣，只是被作者开篇的字眼吸引进来的，所以就想再深入了解一下 Rust ，在此也要先感谢作者的无私！

**Comment by oldoldcoder** [comment]
为什么这样写：
        println!("{}", region);

而不是：
        println!("{}", &region);
貌似已经是切片引用了，为什么还要再引用一下

**Comment by shuyanzi** [comment]
'就像优秀的“国外”课本一样'😂

**Comment by xiamuguizhi** [comment]
@chatgtp4

哈哈，好嘞，咱们一起来幽默解读一下这段代码，绝对不会让你看得云里雾里，带你轻松入门！

首先，你看到的这段代码是用Rust写的，Rust可是编程界的硬核选手，但我们今天要用“软萌”的方式讲解它！

### 1. 企鹅数据，你好！
代码一开始，定义了一个叫 `penguin_data` 的字符串变量，里面保存了一堆企鹅的名字和它们的长度。不过，有个数据叫 `Invalid,data`，看上去就是“捣乱的”，咱们后面会好好对付它。  

```rust
let penguin_data = "\
    common name,length (cm)
    Little penguin,33
    Yellow-eyed penguin,65
    Fiordland penguin,60
    Invalid,data
";
```
这就像我们打开了一张表格，第一行是表头，后面跟着的是各类企鹅的“身高数据”。注意，`Invalid,data` 的存在是个问题，但聪明如我们，马上就能处理。

### 2. 开始遍历企鹅数据
接着，代码用 `.lines()` 把数据按行拆分，每一行代表一个记录。我们用 `for` 循环遍历每一行，同时用 `enumerate()` 给每行配上序号（就像每个企鹅有了排队号码）。

```rust
let records = penguin_data.lines();
 
for (i, record) in records.enumerate() {
```
- **i**：表示这是第几行
- **record**：表示这一行的数据

**搞笑小插曲**：第一行是表头嘛，我们程序员懒得处理这种表头的数据，所以用 `continue` 直接跳过。表头大哥，你先站一边！

```rust
if i == 0 || record.trim().len() == 0 {
    continue;
}
```

### 3. 让数据乖乖“排排坐”
接下来最关键的部分，代码把每行数据按逗号 `,` 分割，然后用 `.trim()` 清理一下两边的空格。`Vec<_>` 这里的 `<_>` 是个小聪明，它告诉Rust：“别问我啥类型，自己搞定！” Rust这点真体贴，不用你操心太多。

```rust
let fields: Vec<_> = record
    .split(',')
    .map(|field| field.trim())
    .collect();
```

**解说员点评**：这样一行数据就变成了一个“字段”列表，比如 `["Little penguin", "33"]`。如果你觉得这个过程有点玄乎，那你可以在调试模式下看到所有数据拆分的细节。

```rust
if cfg!(debug_assertions) {
    eprintln!("debug: {:?} -> {:?}", record, fields);
}
```
这是为我们这些调皮的开发者准备的“作弊”模式，打开它可以让代码输出每次循环的数据。

### 4. 把企鹅身高“量”出来
然后，这段代码会努力地把第二个字段，也就是企鹅的“身高”，转成浮点数 `f32`，如果转换成功，那咱们就开心地输出“这只企鹅有多高”。

```rust
let name = fields[0];
if let Ok(length) = fields[1].parse::<f32>() {
    println!("{}, {}cm", name, length);
}
```
这里用了个 `if let`，有点像一个小型侦探工具。`fields[1]` 里面是身高数字，我们用 `parse::<f32>()` 去试试能不能把它变成数字。

- **成功的情况**：它是个合法数字，比如“33”，那么就会执行`Ok(length)`，把企鹅的名字和身高打出来。
- **失败的情况**：比如“Invalid,data”，那它就会乖乖被忽略，我们装作没看到它。

### 总结：
- **聪明的跳过**：表头这种不重要的东西，直接跳过！
- **灵活的数据处理**：数据有问题？没关系，if let帮你无视那些“捣乱的”！
- **简单的输出**：把合法的企鹅身高输出出来，直接告诉你“小企鹅33厘米高！”。

**Reply by NeoWeb3Nova** [reply]
仿佛看到了AI进化的影子

**Comment by kisun-bit** [comment]
太爱作者了，让我有兴趣往下读

**Comment by xxvv** [comment]
::<f32>冒号冒号看着真别扭

**Comment by Paul-from-NJTech** [comment]
作者大大太 “rust(指的是：秀)” 了

**Comment by utopioooo** [comment]
继续学习。

**Comment by zjwugi** [comment]
good

**Comment by AuxLuffy** [comment]
fields[1].parse::<f32>()
这一行如果数据不规范（没有"," 的时候）会数组越界吧

**Comment by chasays** [comment]
这个用法 太精简了。
if let Ok(length) = fields[1].parse::<f32>() 

返回是的Ok(f32)  ,parse返回的类型正确值， 如果失败了，就是返回其他值，， 然后把f32复制给length

**Comment by yxzzlwj** [comment]
刚hello world突然看到一大串代码，眼前一黑

**Comment by leoiris123** [comment]
我以前入门过rust 这是再次入门了 哈哈. 作者很强很棒很好很幽默

**Comment by langziyang** [comment]
println!("{}", &region); 为什么要用&

**Comment by alansun2** [comment]
现在可以直接循环了，不需要 .iter()，这样也是可以的
```
    for region in regions {
        println!("{}", region);
    }
```

**Reply by NeoWeb3Nova** [reply]
学到后面你就明白， 两种写法会对regions的是否后面的代码使用有决定性意义。
一个是move，一个是借用。

**Comment by LightningJie** [comment]
太帅了，我爱你作者

**Comment by surges** [comment]
很好

**Comment by qiankun25** [comment]
谁懂右上角“繁星点点尽在你的指尖”触发的笑点啊

**Reply by NeoWeb3Nova** [reply]
哈哈，程序员的指尖不是点击粘贴/复制？

**Comment by runqing1** [comment]
刚学hello world就让新手看这种代码，在什么都没有学的前提下就试图讲什么控制流、方法语法、高阶函数、条件编译、隐式返回这些乱七八糟的，而且用词晦涩难懂，净扯什么专有名词，生怕天下人都不知道你会rust一样，这是一个新手教程的大忌！

**Comment by Antecer** [comment]
作为一个稍微有点基础的人，第一段代码就让我发现了新大陆(反而是下面的第二个例子容易理解)，差点没绕出来：
fn greet_world() {
    let chinese = "世界，你好";
    let english = "World, Hello";
    let regions = [chinese, english];
    for region in regions.iter() { println!("{}", &region); }
    for region in &regions { println!("{}", &region); }
    for region in regions { println!("{}", &region); }
    for &region in regions.iter() { println!("{}", region); }
    for &region in &regions { println!("{}", region); }
    for region in regions.iter() { println!("{}", &region); }
    for region in &regions { println!("{}", &region); }
    for region in regions { println!("{}", &region); }
}
